core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,342 @@
+        $this->dispatcher->dispatch($menuEvent, ContaoCoreEvents::FRONTEND_MENU_BUILD);
+
+        return $root;
+    }
+
+    private function getPages(int $pid, array $options): ?array
+    {
+        if (0 === $pid && $options['pages']) {
+            return $this->findPagesByIds($options['pages']);
+        }
+
+        return $this->findPagesByPid($pid, (bool) $options['showHidden']);
+    }
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>|null
+     */
+    private function findPagesByIds(array $pageIds): ?array
+    {
+        // Get all active pages and also include root pages if the language is added to the URL (see #72)
+        $pages = PageModel::findPublishedRegularByIds($pageIds, ['includeRoot' => true]);
+
+        if (null === $pages) {
+            return null;
+        }
+
+        return array_map(
+            static fn (PageModel $page): array => [
+                'page' => $page,
+                'hasSubpages' => false,
+            ],
+            iterator_to_array($pages)
+        );
+    }
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>|null
+     */
+    private function findPagesByPid(int $pid, bool $showHidden = false, bool $isSitemap = false): ?array
+    {
+        $time = Date::floorToMinute();
+        $blnBeUserLoggedIn = $this->tokenChecker->hasBackendUser() && $this->tokenChecker->isPreviewMode();
+        $unroutableTypes = $this->pageRegistry->getUnroutableTypes();
+
+        $pages = $this->connection
+            ->executeQuery("SELECT p1.id, EXISTS(SELECT * FROM tl_page p2 WHERE p2.pid=p1.id AND p2.type!='root' AND p2.type NOT IN ('".implode("', '", $unroutableTypes)."')".(!$showHidden ? ($isSitemap ? " AND (p2.hide='' OR sitemap='map_always')" : " AND p2.hide=''") : '').(!$blnBeUserLoggedIn ? " AND p2.published='1' AND (p2.start='' OR p2.start<='$time') AND (p2.stop='' OR p2.stop>'$time')" : '').") AS hasSubpages FROM tl_page p1 WHERE p1.pid=:pid AND p1.type!='root' AND p1.type NOT IN ('".implode("', '", $unroutableTypes)."')".(!$showHidden ? ($isSitemap ? " AND (p1.hide='' OR sitemap='map_always')" : " AND p1.hide=''") : '').(!$blnBeUserLoggedIn ? " AND p1.published='1' AND (p1.start='' OR p1.start<='$time') AND (p1.stop='' OR p1.stop>'$time')" : '').' ORDER BY p1.sorting', ['pid' => $pid])
+            ->fetchAllAssociative()
+        ;
+
+        if (\count($pages) < 1) {
+            return null;
+        }
+
+        // Load models into the registry with a single query
+        PageModel::findMultipleByIds(array_map(static fn ($row) => $row['id'], $pages));
+
+        return array_map(
+            static fn (array $row): array => [
+                'page' => PageModel::findByPk($row['id']),
+                'hasSubpages' => (bool) $row['hasSubpages'],
+            ],
+            $pages
+        );
+    }
+
+    private function generateUri(PageModel $pageModel, ItemInterface $menuItem): ?string
+    {
+        if ('redirect' === $pageModel->type) {
```

Comment:
Unfortunately no, the router can only generate a URL based on the route, and that can never be the actual target URL outside the page structure. It would generate a link to that page (alias) which then forwards, which is certainly not what we want. I tried to fix this in https://github.com/contao/contao/pull/2031/files but failed so far, I noticed recently that cmf-routing actually has something similar but haven't come up with an implementation yet.